File: app/src/main/java/com/hermex/app/ui/chat/slash/SlashCommandCatalog.kt
package com.hermex.app.ui.chat.slash

import java.util.Locale

/** Android port of the iOS mobile-safe slash-command catalog. */
data class SlashCommand(
    val name: String,
    val description: String,
    val argHint: String? = null,
    val noEcho: Boolean = false,
    val handler: SlashCommandHandler = SlashCommandHandler.Unsupported,
    val subArgs: SlashCommandSubArgs = SlashCommandSubArgs.None
)

sealed interface SlashCommandHandler {
    data object Unsupported : SlashCommandHandler
    data class ClientSide(val action: ClientSideAction) : SlashCommandHandler
    data class ServerSide(val action: ServerSideAction) : SlashCommandHandler
}

enum class ClientSideAction { Clear, Stop, New, Help }

enum class ServerSideAction {
    Model,
    Workspace,
    Reasoning,
    Title,
    Personality,
    Skills,
    Compress,
    Retry,
    Undo,
    Branch,
    Queue,
    Steer,
    Interrupt,
    Status,
    Btw,
    Background,
    Goal
}

enum class SlashCommandSubArgs { None, Models, Personalities, ReasoningLevels, Workspaces, Skills, GoalActions }

data class ParsedSlashQuery(val query: String) {
    private val normalized = query.trimStart()
    val isSlashQuery: Boolean = normalized.startsWith("/")
    private val withoutSlash = if (isSlashQuery) normalized.drop(1) else normalized
    private val firstSpaceIndex = withoutSlash.indexOfFirst { it.isWhitespace() }
    val commandName: String = when {
        !isSlashQuery -> ""
        firstSpaceIndex < 0 -> withoutSlash.trim().lowercase(Locale.ROOT)
        else -> withoutSlash.substring(0, firstSpaceIndex).trim().lowercase(Locale.ROOT)
    }
    val argQuery: String = when {
        !isSlashQuery || firstSpaceIndex < 0 -> ""
        else -> withoutSlash.substring(firstSpaceIndex + 1).trimStart()
    }
    val isSubArgMode: Boolean = isSlashQuery && firstSpaceIndex >= 0
    val command: SlashCommand? = SlashCommandCatalog.command(commandName)
}

object SlashCommandCatalog {
    val reasoningLevels = listOf("show", "hide", "none", "minimal", "low", "medium", "high", "xhigh")
    val goalActions = listOf("status", "pause", "resume", "clear")

    val allCommands: List<SlashCommand> = listOf(
        SlashCommand("help", "Show available slash commands", handler = SlashCommandHandler.ClientSide(ClientSideAction.Help)),
        SlashCommand("clear", "Clear the current conversation", noEcho = true, handler = SlashCommandHandler.ClientSide(ClientSideAction.Clear)),
        SlashCommand("model", "Switch the active model", "model_name", true, SlashCommandHandler.ServerSide(ServerSideAction.Model), SlashCommandSubArgs.Models),
        SlashCommand("workspace", "Switch the active workspace", "path", true, SlashCommandHandler.ServerSide(ServerSideAction.Workspace), SlashCommandSubArgs.Workspaces),
        SlashCommand("reasoning", "Set reasoning effort level", "level", true, SlashCommandHandler.ServerSide(ServerSideAction.Reasoning), SlashCommandSubArgs.ReasoningLevels),
        SlashCommand("new", "Start a new session", noEcho = true, handler = SlashCommandHandler.ClientSide(ClientSideAction.New)),
        SlashCommand("stop", "Stop the current response", noEcho = true, handler = SlashCommandHandler.ClientSide(ClientSideAction.Stop)),
        SlashCommand("title", "Rename the current session", "name", false, SlashCommandHandler.ServerSide(ServerSideAction.Title)),
        SlashCommand("personality", "Set the session personality", "name", false, SlashCommandHandler.ServerSide(ServerSideAction.Personality), SlashCommandSubArgs.Personalities),
        SlashCommand("skills", "Search available skills", "query", false, SlashCommandHandler.ServerSide(ServerSideAction.Skills), SlashCommandSubArgs.Skills),
        SlashCommand("compress", "Compress session context", "focus topic", true, SlashCommandHandler.ServerSide(ServerSideAction.Compress)),
        SlashCommand("compact", "Alias for /compress", "focus topic", true, SlashCommandHandler.ServerSide(ServerSideAction.Compress)),
        SlashCommand("retry", "Retry the last turn", noEcho = true, handler = SlashCommandHandler.ServerSide(ServerSideAction.Retry)),
        SlashCommand("undo", "Undo the last exchange", noEcho = true, handler = SlashCommandHandler.ServerSide(ServerSideAction.Undo)),
        SlashCommand("branch", "Fork the conversation", "name", true, SlashCommandHandler.ServerSide(ServerSideAction.Branch)),
        SlashCommand("fork", "Alias for /branch", "name", true, SlashCommandHandler.ServerSide(ServerSideAction.Branch)),
        SlashCommand("queue", "Queue a message for the next turn", "message", true, SlashCommandHandler.ServerSide(ServerSideAction.Queue)),
        SlashCommand("steer", "Steer the active response", "message", true, SlashCommandHandler.ServerSide(ServerSideAction.Steer)),
        SlashCommand("interrupt", "Stop the response and send a new message", "message", true, SlashCommandHandler.ServerSide(ServerSideAction.Interrupt)),
        SlashCommand("status", "Show session status", handler = SlashCommandHandler.ServerSide(ServerSideAction.Status)),
        SlashCommand("goal", "Set or inspect a persistent goal", "[status|pause|resume|clear|text]", true, SlashCommandHandler.ServerSide(ServerSideAction.Goal), SlashCommandSubArgs.GoalActions),
        SlashCommand("btw", "Ask a side question", "question", true, SlashCommandHandler.ServerSide(ServerSideAction.Btw)),
        SlashCommand("background", "Run a parallel task", "prompt", true, SlashCommandHandler.ServerSide(ServerSideAction.Background)),
        SlashCommand("bg", "Alias for /background", "prompt", true, SlashCommandHandler.ServerSide(ServerSideAction.Background))
    )

    fun matching(query: String): List<SlashCommand> {
        val trimmed = query.trim()
        if (trimmed.isEmpty()) return allCommands
        val lower = trimmed.lowercase(Locale.ROOT)
        return allCommands.filter {
            it.name.lowercase(Locale.ROOT).startsWith(lower) ||
                it.description.lowercase(Locale.ROOT).contains(lower)
        }
    }

    fun command(name: String): SlashCommand? {
        val lower = name.lowercase(Locale.ROOT)
        return allCommands.firstOrNull { it.name.lowercase(Locale.ROOT) == lower }
    }
}

File: HermesMobile/Features/Chat/SlashCommandCatalog.swift
import Foundation

enum SlashCommandCatalog {
    static let allCommands: [SlashCommand] = [
        SlashCommand(
            name: "help",
            description: String(localized: "Show available slash commands"),
            noEcho: false,
            handler: .clientSide(.help)
        ),
        SlashCommand(
            name: "clear",
            description: String(localized: "Clear this conversation on the server"),
            noEcho: true,
            handler: .clientSide(.clear)
        ),
        SlashCommand(
            name: "model",
            description: String(localized: "Switch the active model"),
            argHint: String(localized: "model_name"),
            noEcho: true,
            handler: .serverSide(.model),
            subArgs: .models
        ),
        SlashCommand(
            name: "workspace",
            description: String(localized: "Switch the active workspace"),
            argHint: String(localized: "path"),
            noEcho: true,
            handler: .serverSide(.workspace),
            subArgs: .workspaces
        ),
        SlashCommand(
            name: "reasoning",
            description: String(localized: "Set reasoning effort level"),
            argHint: String(localized: "level"),
            noEcho: true,
            handler: .serverSide(.reasoning),
            subArgs: .reasoningLevels
        ),
        SlashCommand(
            name: "new",
            description: String(localized: "Start a new session"),
            noEcho: true,
            handler: .clientSide(.new)
        ),
        SlashCommand(
            name: "stop",
            description: String(localized: "Stop the current response"),
            noEcho: true,
            handler: .clientSide(.stop)
        ),
        SlashCommand(
            name: "title",
            description: String(localized: "Rename the current session"),
            argHint: String(localized: "name"),
            noEcho: false,
            handler: .serverSide(.title)
        ),
        SlashCommand(
            name: "personality",
            description: String(localized: "Set the session personality"),
            argHint: String(localized: "name"),
            noEcho: false,
            handler: .serverSide(.personality),
            subArgs: .personalities
        ),
        SlashCommand(
            name: "skills",
            description: String(localized: "Search available skills"),
            argHint: String(localized: "query"),
            noEcho: false,
            handler: .serverSide(.skills),
            subArgs: .skills
        ),
        SlashCommand(
            name: "compress",
            description: String(localized: "Compress session context"),
            argHint: String(localized: "focus topic"),
            noEcho: true,
            handler: .serverSide(.compress)
        ),
        SlashCommand(
            name: "compact",
            description: String(localized: "Alias for \("/compress")"),
            argHint: String(localized: "focus topic"),
            noEcho: true,
            handler: .serverSide(.compress)
        ),
        SlashCommand(
            name: "retry",
            description: String(localized: "Retry the last turn"),
            noEcho: true,
            handler: .serverSide(.retry)
        ),
        SlashCommand(
            name: "undo",
            description: String(localized: "Undo the last exchange"),
            noEcho: true,
            handler: .serverSide(.undo)
        ),
        SlashCommand(
            name: "branch",
            description: String(localized: "Fork the conversation"),
            argHint: String(localized: "name"),
            noEcho: true,
            handler: .serverSide(.branch)
        ),
        SlashCommand(
            name: "fork",
            description: String(localized: "Alias for \("/branch")"),
            argHint: String(localized: "name"),
            noEcho: true,
            handler: .serverSide(.branch)
        ),
        SlashCommand(
            name: "queue",
            description: String(localized: "Queue a message for the next turn"),
            argHint: String(localized: "message"),
            noEcho: true,
            handler: .serverSide(.queue)
        ),
        SlashCommand(
            name: "steer",
            description: String(localized: "Steer the active response"),
            argHint: String(localized: "message"),
            noEcho: true,
            handler: .serverSide(.steer)
        ),
        SlashCommand(
            name: "interrupt",
            description: String(localized: "Stop the response and send a new message"),
            argHint: String(localized: "message"),
            noEcho: true,
            handler: .serverSide(.interrupt)
        ),
        SlashCommand(
            name: "status",
            description: String(localized: "Show session status"),
            noEcho: false,
            handler: .serverSide(.status)
        ),
        SlashCommand(
            name: "goal",
            description: String(localized: "Set or inspect a persistent goal"),
            argHint: "[status|pause|resume|clear|text]",
            noEcho: true,
            handler: .serverSide(.goal),
            subArgs: .goalActions
        ),
        SlashCommand(
            name: "btw",
            description: String(localized: "Ask a side question"),
            argHint: String(localized: "question"),
            noEcho: true,
            handler: .serverSide(.btw)
        ),
        SlashCommand(
            name: "background",
            description: String(localized: "Run a parallel task"),
            argHint: String(localized: "prompt"),
            noEcho: true,
            handler: .serverSide(.background)
        ),
        SlashCommand(
            name: "bg",
            description: String(localized: "Alias for \("/background")"),
            argHint: String(localized: "prompt"),
            noEcho: true,
            handler: .serverSide(.background)
        )
    ]

    /// The commands worth showing for `query`, best first. An empty query keeps
    /// the curated order above.
    static func matching(_ query: String, limit: Int = SlashCommandRanker.resultLimit) -> [SlashCommand] {
        SlashCommandRanker.rank(allCommands, matching: query, limit: limit) { command in
            SlashRankableFields(name: command.name, description: command.description)
        }
    }

    /// Every built-in command name, lowercased. Used to keep an agent command
    /// that shadows a built-in out of the popover.
    static let builtinNames: Set<String> = Set(allCommands.map { $0.name.lowercased() })

    static func command(named name: String) -> SlashCommand? {
        allCommands.first { $0.name.lowercased() == name.lowercased() }
    }

    static let reasoningLevels = ["show", "hide", "none", "minimal", "low", "medium", "high", "xhigh"]
    static let goalActions = ["status", "pause", "resume", "clear"]
}
